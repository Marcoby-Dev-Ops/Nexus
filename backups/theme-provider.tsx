import React, { createContext, useContext, useEffect, useState } from 'react';

type Theme = 'dark' | 'light' | 'system';

interface ThemeProviderProps {
  children: React.ReactNode;
  defaultTheme?: Theme;
  storageKey?: string;
}

interface ThemeProviderState {
  theme: Theme;
  setTheme: (theme: Theme) => void;
}

const initialState: ThemeProviderState = {
  theme: 'light',
  setTheme: () => null,
};

const ThemeProviderContext = createContext<ThemeProviderState>(initialState);

/**
 * @name ThemeProvider
 * @description Provides theme context to the application for light/dark/system mode.
 * @param {ThemeProviderProps} props - The props for the component.
 * @returns {JSX.Element}
 */
export function ThemeProvider({ 
  children, 
  defaultTheme = 'light', 
  storageKey = 'vite-ui-theme', 
  ...props 
}: ThemeProviderProps) {
  const [theme, setTheme] = useState<Theme>(
    () => (localStorage.getItem(storageKey) as Theme) || defaultTheme
  );

  useEffect(() => {
    const root = window.document.documentElement;
    root.classList.remove('light', 'dark');

    if (theme === 'system') {
      const systemTheme = window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light';
      root.classList.add(systemTheme);
      return;
    }

    root.classList.add(theme);
  }, [theme]);

  const value = {
    theme,
    setTheme: (newTheme: Theme) => {
      localStorage.setItem(storageKey, newTheme);
      setTheme(newTheme);
    },
  };

  return (
    <ThemeProviderContext.Provider {...props} value={value}>
      {children}
    </ThemeProviderContext.Provider>
  );
}

/**
 * @name useTheme
 * @description Hook to access the current theme and a function to update it.
 * @returns {ThemeProviderState} The current theme state and setter.
 */
export const useTheme = () => {
  const context = useContext(ThemeProviderContext);
  if (context === undefined) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
}; 