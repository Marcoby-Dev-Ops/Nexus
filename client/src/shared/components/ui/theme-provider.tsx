import React, { createContext, useContext, useEffect, useState } from 'react';
import { safeGetLocalStorage, safeSetLocalStorage } from '@/shared/utils/storageUtils';
import { THEME_COLORS, type Theme, type PrimaryColor } from '@/shared/constants/theme';
import { logger } from '@/shared/utils/logger';

type ThemeProviderProps = {
  children: React.ReactNode;
  defaultTheme?: Theme;
  defaultColor?: PrimaryColor;
};

type ThemeProviderState = {
  theme: Theme;
  setTheme: (theme: Theme) => void;
  primaryColor: PrimaryColor;
  setPrimaryColor: (color: PrimaryColor) => void;
};

const initialState: ThemeProviderState = {
  theme: 'system',
  setTheme: () => null,
  primaryColor: 'green',
  setPrimaryColor: () => null,
};



const ThemeProviderContext = createContext<ThemeProviderState>(initialState);

/**
 * @name ThemeProvider
 * @description Provides theme context to the application for light/dark/system mode.
 * @param {ThemeProviderProps} props - The props for the component.
 * @returns {JSX.Element}
 */
export function ThemeProvider({ 
  children, 
  defaultTheme = 'system', 
  defaultColor = 'green',
  ...props 
}: ThemeProviderProps) {
  // Prevent duplicate "initial" logs under React StrictMode double render in dev
  // Module-level flags ensure we only log the initial load once per window
  type WindowWithNexusFlags = Window & {
    __nexus_hasLoggedInitialTheme?: boolean;
    __nexus_hasLoggedInitialPrimary?: boolean;
    __nexus_themeApplied?: boolean;
    __nexus_primaryApplied?: boolean;
  };
  const w = window as WindowWithNexusFlags;

  const [theme, setTheme] = useState<Theme>(() => {
    try {
      const storedTheme = safeGetLocalStorage<Theme>('theme', defaultTheme);
      if (!w.__nexus_hasLoggedInitialTheme) {
        logger.info('[ThemeProvider] Initial theme loaded:', storedTheme, 'defaultTheme:', defaultTheme);
        w.__nexus_hasLoggedInitialTheme = true;
      }
      return storedTheme;
    } catch (error) {
      logger.error('[ThemeProvider] Error loading theme from localStorage:', error);
      return defaultTheme;
    }
  });
  
  const [primaryColor, setPrimaryColor] = useState<PrimaryColor>(() => {
    try {
      const storedColor = safeGetLocalStorage<PrimaryColor>('primaryColor', defaultColor);
      if (!w.__nexus_hasLoggedInitialPrimary) {
        logger.info('[ThemeProvider] Initial primary color loaded:', storedColor);
        w.__nexus_hasLoggedInitialPrimary = true;
      }
      return storedColor;
    } catch (error) {
      logger.error('[ThemeProvider] Error loading primary color from localStorage:', error);
      return defaultColor;
    }
  });

  // Apply theme with better React StrictMode handling
  useEffect(() => {
    try {
      const root = window.document.documentElement;
      root.classList.remove('light', 'dark');

      if (theme === 'system') {
        const systemTheme = window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light';
        root.classList.add(systemTheme);
        // Only log if not already applied to prevent duplicate logs in StrictMode
        if (!w.__nexus_themeApplied) {
          logger.info('[ThemeProvider] Applied system theme:', systemTheme);
          w.__nexus_themeApplied = true;
        }
      } else {
        root.classList.add(theme);
        // Only log if not already applied to prevent duplicate logs in StrictMode
        if (!w.__nexus_themeApplied) {
          logger.info('[ThemeProvider] Applied theme:', theme);
          w.__nexus_themeApplied = true;
        }
      }
    } catch (error) {
      logger.error('[ThemeProvider] Error applying theme:', error);
    }
  }, [theme]);

  // Apply primary color with better React StrictMode handling
  useEffect(() => {
    try {
      const root = window.document.documentElement;
      const color = THEME_COLORS.find(c => c.name === primaryColor);
      if (color) {
        root.style.setProperty('--primary', color.value);
        root.style.setProperty('--primary-foreground', color.foreground);
        // Only log if not already applied to prevent duplicate logs in StrictMode
        if (!w.__nexus_primaryApplied) {
          logger.info('[ThemeProvider] Applied primary color:', primaryColor);
          w.__nexus_primaryApplied = true;
        }
      }
    } catch (error) {
      logger.error('[ThemeProvider] Error applying primary color:', error);
    }
  }, [primaryColor]);

  const value = {
    theme,
    setTheme: (theme: Theme) => {
      try {
        safeSetLocalStorage('theme', theme);
        setTheme(theme);
        logger.info('[ThemeProvider] Theme updated:', theme);
      } catch (error) {
        logger.error('[ThemeProvider] Error saving theme to localStorage:', error);
        setTheme(theme); // Still update the state even if localStorage fails
      }
    },
    primaryColor,
    setPrimaryColor: (color: PrimaryColor) => {
      try {
        safeSetLocalStorage('primaryColor', color);
        setPrimaryColor(color);
        logger.info('[ThemeProvider] Primary color updated:', color);
      } catch (error) {
        logger.error('[ThemeProvider] Error saving primary color to localStorage:', error);
        setPrimaryColor(color); // Still update the state even if localStorage fails
      }
    },
  };

  return (
    <ThemeProviderContext.Provider {...props} value={value}>
      {children}
    </ThemeProviderContext.Provider>
  );
}

/**
 * @name useTheme
 * @description Hook to access the current theme and a function to update it.
 * @returns {ThemeProviderState} The current theme state and setter.
 */
export const useTheme = () => {
  const context = useContext(ThemeProviderContext);
  if (context === undefined) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
}; 
