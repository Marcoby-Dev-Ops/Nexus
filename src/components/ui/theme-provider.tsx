import React, { createContext, useContext, useEffect, useState } from 'react';
import { safeGetLocalStorage, safeSetLocalStorage } from '../../lib/utils/storageUtils';

type Theme = 'dark' | 'light' | 'system';

type ThemeProviderProps = {
  children: React.ReactNode;
  defaultTheme?: Theme;
  defaultColor?: string;
};

type ThemeProviderState = {
  theme: Theme;
  setTheme: (theme: Theme) => void;
  primaryColor: string;
  setPrimaryColor: (color: string) => void;
};

const initialState: ThemeProviderState = {
  theme: 'system',
  setTheme: () => null,
  primaryColor: 'blue',
  setPrimaryColor: () => null,
};

export const COLORS = [
  { name: 'blue', value: '221.2 83.2% 53.3%', foreground: '210 40% 98%' },
  { name: 'green', value: '142.1 76.2% 36.3%', foreground: '210 40% 98%' },
  { name: 'orange', value: '24.6 95% 53.1%', foreground: '210 40% 98%' },
  { name: 'red', value: '0 84.2% 60.2%', foreground: '210 40% 98%' },
  { name: 'purple', value: '262.1 83.3% 57.8%', foreground: '210 40% 98%' },
];

const ThemeProviderContext = createContext<ThemeProviderState>(initialState);

/**
 * @name ThemeProvider
 * @description Provides theme context to the application for light/dark/system mode.
 * @param {ThemeProviderProps} props - The props for the component.
 * @returns {JSX.Element}
 */
export function ThemeProvider({ 
  children, 
  defaultTheme = 'system', 
  defaultColor = 'blue',
  ...props 
}: ThemeProviderProps) {
  const [theme, setTheme] = useState<Theme>(
    () => safeGetLocalStorage<Theme>('theme', defaultTheme)
  );
  const [primaryColor, setPrimaryColor] = useState<string>(
    () => safeGetLocalStorage<string>('primaryColor', defaultColor)
  );

  useEffect(() => {
    const root = window.document.documentElement;
    root.classList.remove('light', 'dark');

    if (theme === 'system') {
      const systemTheme = window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light';
      root.classList.add(systemTheme);
    } else {
      root.classList.add(theme);
    }
  }, [theme]);

  useEffect(() => {
    const root = window.document.documentElement;
    const color = COLORS.find(c => c.name === primaryColor);
    if (color) {
      root.style.setProperty('--primary', color.value);
      root.style.setProperty('--primary-foreground', color.foreground);
    }
  }, [primaryColor]);

  const value = {
    theme,
    setTheme: (theme: Theme) => {
      safeSetLocalStorage('theme', theme);
      setTheme(theme);
    },
    primaryColor,
    setPrimaryColor: (color: string) => {
      safeSetLocalStorage('primaryColor', color);
      setPrimaryColor(color);
    },
  };

  return (
    <ThemeProviderContext.Provider {...props} value={value}>
      {children}
    </ThemeProviderContext.Provider>
  );
}

/**
 * @name useTheme
 * @description Hook to access the current theme and a function to update it.
 * @returns {ThemeProviderState} The current theme state and setter.
 */
export const useTheme = () => {
  const context = useContext(ThemeProviderContext);
  if (context === undefined) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
}; 