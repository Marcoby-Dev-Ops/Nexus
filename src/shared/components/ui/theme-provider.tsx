import React, { createContext, useContext, useEffect, useState } from 'react';
import { safeGetLocalStorage, safeSetLocalStorage } from '@/shared/utils/storageUtils.ts';

type Theme = 'dark' | 'light' | 'system';

type ThemeProviderProps = {
  children: React.ReactNode;
  defaultTheme?: Theme;
  defaultColor?: string;
};

type ThemeProviderState = {
  theme: Theme;
  setTheme: (theme: Theme) => void;
  primaryColor: string;
  setPrimaryColor: (color: string) => void;
};

const initialState: ThemeProviderState = {
  theme: 'system',
  setTheme: () => null,
  primaryColor: 'green',
  setPrimaryColor: () => null,
};

export const COLORS = [
  { name: 'blue', value: '221.2 83.2% 53.3%', foreground: '210 40% 98%' },
  { name: 'green', value: '142.1 76.2% 36.3%', foreground: '210 40% 98%' },
  { name: 'orange', value: '24.6 95% 53.1%', foreground: '210 40% 98%' },
  { name: 'red', value: '0 84.2% 60.2%', foreground: '210 40% 98%' },
  { name: 'purple', value: '262.1 83.3% 57.8%', foreground: '210 40% 98%' },
];

const ThemeProviderContext = createContext<ThemeProviderState>(initialState);

/**
 * @name ThemeProvider
 * @description Provides theme context to the application for light/dark/system mode.
 * @param {ThemeProviderProps} props - The props for the component.
 * @returns {JSX.Element}
 */
export function ThemeProvider({ 
  children, 
  defaultTheme = 'system', 
  defaultColor = 'green',
  ...props 
}: ThemeProviderProps) {
  const [theme, setTheme] = useState<Theme>(() => {
    try {
      const storedTheme = safeGetLocalStorage<Theme>('theme', defaultTheme);
      console.log('[ThemeProvider] Initial theme loaded:', storedTheme, 'defaultTheme:', defaultTheme);
      return storedTheme;
    } catch (error) {
      console.error('[ThemeProvider] Error loading theme from localStorage:', error);
      return defaultTheme;
    }
  });
  
  const [primaryColor, setPrimaryColor] = useState<string>(() => {
    try {
      const storedColor = safeGetLocalStorage<string>('primaryColor', defaultColor);
      console.log('[ThemeProvider] Initial primary color loaded:', storedColor);
      return storedColor;
    } catch (error) {
      console.error('[ThemeProvider] Error loading primary color from localStorage:', error);
      return defaultColor;
    }
  });

  useEffect(() => {
    try {
      const root = window.document.documentElement;
      root.classList.remove('light', 'dark');

      if (theme === 'system') {
        const systemTheme = window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light';
        root.classList.add(systemTheme);
        console.log('[ThemeProvider] Applied system theme:', systemTheme);
      } else {
        root.classList.add(theme);
        console.log('[ThemeProvider] Applied theme:', theme);
      }
    } catch (error) {
      console.error('[ThemeProvider] Error applying theme:', error);
    }
  }, [theme]);

  useEffect(() => {
    try {
      const root = window.document.documentElement;
      const color = COLORS.find(c => c.name === primaryColor);
      if (color) {
        root.style.setProperty('--primary', color.value);
        root.style.setProperty('--primary-foreground', color.foreground);
        console.log('[ThemeProvider] Applied primary color:', primaryColor);
      }
    } catch (error) {
      console.error('[ThemeProvider] Error applying primary color:', error);
    }
  }, [primaryColor]);

  const value = {
    theme,
    setTheme: (theme: Theme) => {
      try {
        safeSetLocalStorage('theme', theme);
        setTheme(theme);
        console.log('[ThemeProvider] Theme updated:', theme);
      } catch (error) {
        console.error('[ThemeProvider] Error saving theme to localStorage:', error);
        setTheme(theme); // Still update the state even if localStorage fails
      }
    },
    primaryColor,
    setPrimaryColor: (color: string) => {
      try {
        safeSetLocalStorage('primaryColor', color);
        setPrimaryColor(color);
        console.log('[ThemeProvider] Primary color updated:', color);
      } catch (error) {
        console.error('[ThemeProvider] Error saving primary color to localStorage:', error);
        setPrimaryColor(color); // Still update the state even if localStorage fails
      }
    },
  };

  return (
    <ThemeProviderContext.Provider {...props} value={value}>
      {children}
    </ThemeProviderContext.Provider>
  );
}

/**
 * @name useTheme
 * @description Hook to access the current theme and a function to update it.
 * @returns {ThemeProviderState} The current theme state and setter.
 */
export const useTheme = () => {
  const context = useContext(ThemeProviderContext);
  if (context === undefined) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
}; 