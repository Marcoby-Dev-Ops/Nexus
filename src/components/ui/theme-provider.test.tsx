/**
 * @file theme-provider.test.tsx
 * @description Unit and snapshot tests for the theme-provider component.
 */
import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { ThemeProvider } from './theme-provider';

describe('ThemeProvider', () => {
  it('renders children', () => {
    render(
      <ThemeProvider>
        <div>Theme Content</div>
      </ThemeProvider>
    );
    expect(screen.getByText('Theme Content')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { asFragment } = render(
      <ThemeProvider>
        <div>Snapshot Theme</div>
      </ThemeProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
}); 