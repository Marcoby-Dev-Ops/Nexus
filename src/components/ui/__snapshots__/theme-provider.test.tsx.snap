// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider matches snapshot 1`] = `
<DocumentFragment>
  <div>
    Snapshot Theme
  </div>
</DocumentFragment>
`;
